IBM/carbon-addons-iot-react
# fix(table): column resizing doesn't work after column reordering

Closes #1016

**Summary**
Solves two issues originally described in internal [312](https://github.ibm.com/Watson-IoT/pal-tracking/issues/312)  

1. Moved Columns
When you drag to re-order columns and then attempt to re-size, the table seems to "remember" where the columns originally were, and takes width from that area, which feels different than the intended UX.

2. Hidden columns
When you hide columns, the right-hand side of the table is still considered occupied, and you cannot drag the furthest-right column to fill the space.
(This is issue is more complicated and have corner cases not solved with this fix, they are tracked here [1019](https://github.com/IBM/carbon-addons-iot-react/issues/1019) )

**Change List (commits, features, bugs, etc)**
- Updated table stories so that the two issues and fixes could be verified 
- Updated ColumnResize.jsx with improved naming and correct identification of affected column sibling
- Created `columnWidthUtilityFunctions.js` for dealing with the calculations of column widths. Previously the widths where measured instead of calculated after column toggle which meant that the row action column kept growing (point 2). Hopefully this strategy is also faster compared to the previous since it doesn't require an extra render.
- Removed the useEffect in TableHead since `onColumnResize(updatedColumns)` now can be called without measuring the new widths.
- Added px to the values exported by the onColumnResize (is that a breaking change?)
- Created unit tests for point 2

**Left Todo**
- Working on creating meaningful unit tests for point 1 and other resize functionality 

**Acceptance Test (how to verify the PR)**
_Point 1_
1. Go to https://deploy-preview-1039--carbon-addons-iot-react.netlify.com/?path=/story/watson-iot-table--with-resize-hascolumnselection-and-initial-column-widths
2. Click the column selection button i the top right corner 
3. Rearrange the columns (this might be slow, there is another PR for that here [1034]( https://github.com/IBM/carbon-addons-iot-react/pull/1034) )
4. Resize the columns and make sure it works as expected
5. Also make sure that the emitted values for actions `onColumnResize` and `onChangeOrdering` are correct 

_Point 2_
1. Go to https://deploy-preview-1039--carbon-addons-iot-react.netlify.com/?path=/story/watson-iot-table--stateful-example-with-expansion-and-column-resize
2. Click the column selection button i the top right corner 
3. Hide a couple of columns and make sure that the action column doesn't grow in width
4. Show the hidden columns and verify that the action column width stays the same
